pkg/cloudprovider/provider/baremetal/plugins/tinkerbell/client/workflow.go

```diff
@@ -18,203 +18,89 @@ package client
-			return nil, fmt.Errorf("workflow %w", ErrNotFound)
-		}
+	tink "github.com/kubermatic/machine-controller/pkg/cloudprovider/provider/baremetal/plugins/tinkerbell/types"
 
-		return nil, fmt.Errorf("getting workflow version from Tinkerbell: %w", err)
-	}
-
-	req := &workflow.GetWorkflowDataRequest{WorkflowId: id, Version: verResp.GetVersion()}
-
-	resp, err := t.client.GetWorkflowMetadata(ctx, req)
-	if err != nil {
-		if err.Error() == sqlErrorString || err.Error() == sqlErrorStringAlt {
-			return nil, fmt.Errorf("workflow %w", ErrNotFound)
-		}
-
-		return nil, fmt.Errorf("getting workflow metadata from Tinkerbell: %w", err)
-	}
+	tinkv1alpha1 "github.com/tinkerbell/tink/api/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
 
-	return resp.GetData(), nil
+// WorkflowClient handles interactions with the Tinkerbell Workflows.
+type WorkflowClient struct {
+	tinkclient client.Client
 }
 
-// GetActions returns the actions for a given Tinkerbell Workflow.
-func (t *Workflow) GetActions(ctx context.Context, id string) ([]*workflow.WorkflowAction, error) {
-	req := &workflow.WorkflowActionsRequest{WorkflowId: id}
-
-	resp, err := t.client.GetWorkflowActions(ctx, req)
-	if err != nil {
-		if err.Error() == sqlErrorString || err.Error() == sqlErrorStringAlt {
-			return nil, fmt.Errorf("workflow %w", ErrNotFound)
-		}
-
-		return nil, fmt.Errorf("getting workflow actions from Tinkerbell: %w", err)
+// NewWorkflowClient creates a new client for managing Tinkerbell workflows.
+func NewWorkflowClient(k8sClient client.Client) *WorkflowClient {
+	return &WorkflowClient{
+		tinkclient: k8sClient,
 	}
-
-	return resp.GetActionList(), nil
 }
 
-// GetEvents returns the events for a given Tinkerbell Workflow.
-func (t *Workflow) GetEvents(ctx context.Context, id string) ([]*workflow.WorkflowActionStatus, error) {
-	req := &workflow.GetRequest{Id: id}
-
-	resp, err := t.client.ShowWorkflowEvents(ctx, req)
-	if err != nil {
-		if err.Error() == sqlErrorString || err.Error() == sqlErrorStringAlt {
-			return nil, fmt.Errorf("workflow %w", ErrNotFound)
-		}
+// CreateWorkflow creates a new Tinkerbell Workflow resource in the cluster.
+func (w *WorkflowClient) CreateWorkflow(ctx context.Context, userData, workflowName, templateRef string, hardware tink.Hardware) error {
+	// Construct the Workflow object
+	ifaceConfig := hardware.Spec.Interfaces[0].DHCP
+	dnsNameservers := "1.1.1.1"
 
-		return nil, fmt.Errorf("getting workflow events from Tinkerbell: %w", err)
+	for _, ns := range ifaceConfig.NameServers {
+		dnsNameservers = ns
 	}
 
-	result := []*workflow.WorkflowActionStatus{}
-
-	for {
-		e, err := resp.Recv()
-		if errors.Is(err, io.EOF) {
-			break
-		}
-
-		if err != nil {
-			return nil, fmt.Errorf("getting workflow event from Tinkerbell: %w", err)
-		}
-
-		result = append(result, e)
+	workflow := &tinkv1alpha1.Workflow{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      workflowName + "-workflow",
```

Comment:
Please somehow make this single line re-usable. At the moment it's both an implementation detail, but also used by the Tinkerbell driver directly (i.e. it has knowledge it should not have).

Also, is it intended that "-workflow" is appended to the _workflow_ name, or is the parameter for this function misnamed? In the driver, the server's name is used for this parameter.

Reply:
Yes we can but honestly I took over this PR only yesterday and I cannot fix everything in one shot 😅 I can address that in up coming PRs